### PR TITLE
Implement `blank_line_after_nested_stub_class` preview style

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR implements the `blank_line_after_nested_stub_class` preview style in the formatter.

The logic is divided into 3 parts:
1. In between preceding and following nodes at top level and nested suite
2. When there's a trailing comment after the class
3. When there is no following node from (1) which is the case when it's the last or the only node in a suite

We handle (3) with `FormatLeadingAlternateBranchComments`.

## Test Plan

- Add new test cases and update existing snapshots
- Checked the `typeshed` diff

fixes: #8891

### CODE REVIEW

File: crates/ruff_python_formatter/tests/snapshots/format@stub_files__suite.pyi.snap [hunk 1]
```diff
@@ -309,4 +309,27 @@ class ComplexStatements:
...
+```diff
+--- Stable
++++ Preview
+@@ -110,6 +110,7 @@
+ 
+     class InnerClass5:
+         def a(self): ...
++
+     field1 = 1
+
```
Comment (dhruvmanila) [inline crates/ruff_python_formatter/tests/snapshots/format@stub_files__suite.pyi.snap:322]: This seems like a bug in Black's implementation: https://github.com/psf/black/issues/4113

File: crates/ruff_python_formatter/src/statement/suite.rs [hunk 2]
```diff
@@ -447,7 +448,11 @@ fn stub_file_empty_lines(
         }
         SuiteKind::Class | SuiteKind::Other | SuiteKind::Function => {
             if empty_line_condition
-                && lines_after_ignoring_end_of_line_trivia(preceding.end(), source) > 1
+                && ((is_blank_line_after_nested_stub_class_enabled(f.context())
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/statement/suite.rs:451]: Should this also apply if the class is nested in a if/while or any other control flow block?

**Review by konstin — APPROVED** [review]

**Review by MichaReiser — APPROVED** [review]

**Comment by MichaReiser** [comment]
@dhruvmanila what's the status of this PR? Also note that there have been some recent changes in Black. It might be worth double checking Black's tests again to ensure we match the formatting precisely.

**Comment by dhruvmanila** [comment]
Here's an update regarding this preview style implementation in Ruff:

tldr; there are bunch of edge cases where it seems to be difficult to add support ...

### What's working today:

#### 1

Cases when the class definition is followed by another statement at the same indent level. This is to say that in a suite, the preceding and following value is set. Why is this easier? Because [the loop] takes care of a lot of conditions.

For example, in the following code snippet there are 2 statements in a suite which is the body of the `Top` class. Here, the preceding would be `Nested1` class and following would be the assignment statement. Because we have the context of the siblings, it's easier to find out if the class needs an empty line or not.

```python
class Top:
	class Nested1:  # <- statement 1
		pass
	foo = 1  # <- statement 2
```

#### 2

Some other cases which works today is where the class is followed by a trailing comment. This is then handled by the class formatting as then it doesn't matter if there's any other statement following the class or not as we'll always need to add an empty line between the class and the comment. For example,

```python
class Top:
	class Nested1:
		pass
	# comment
```

This will be handled in `empty_lines_before_trailing_comments` builder.

### Problematic cases

Note that some of the following cases have a workaround in the PR but I'm not fully convinced with the implementation which is why they've been put in this section.

The main type of case where there are problems are the ones where the class definition is the last statement in the suite. For example,

#### 1

```python
class Top:
	class Nested:
		pass
foo = 1
```

Here, [the loop] will never be executed because there's just 1 statement in the suite. This means that we need to have a special case implementation for the last statement in the suite. This needs to account for a lot of other conditions to make the decision of whether to add an empty line or not:

Is there a class in the parent suite which already added an empty line? If so, then we don't need to add another empty line.

```python
class Top:
    class Nested11:
        class Nested12:
            pass
    class Nested21:
        pass
```

Here, the `Nested11` class will add an empty line. Now, the `Nested12` class should not be adding an empty line otherwise there'll be 2 empty lines. **This is solved in this PR** by thinking it in a different way. Instead of the `Nested11` class being responsible for adding an empty line, we'll make `Nested12` class responsible for it. This is to say that the innermost class should be responsible to add an empty line. Why? Because as a parent it's easier to traverse through the children to check this condition:

```rust
/// Checks if the last child of the given node is a class definition without a
/// trailing comment.
pub(crate) fn is_last_child_a_class_def(node: AnyNodeRef<'_>, f: &PyFormatter) -> bool {
    let comments = f.context().comments();
    std::iter::successors(node.last_child_in_body(), AnyNodeRef::last_child_in_body)
        .take_while(|last_child| !comments.has_trailing_own_line(*last_child))
        .any(|last_child| last_child.is_stmt_class_def())
}
```

But, this then creates another problem. Take the following code as an example:

```python
if a:
	class Nested:
		pass
else:
	pass
```

Here, the problem is that the new logic will add an empty line after the `Nested` class which will give us:

```python
if a:
	class Nested:
		pass

else:
	pass
```

Now, the same logic will try to _preserve_ the empty line by adding the IR element in the same position. But while formatting the `else` block, the `FormatLeadingAlternateBranchComments` formatter sees that there's an empty line before the `else` block which it tries to maintain by adding an empty line IR element. Thus, two empty line elements are added which creates an unstable formatting.

#### 2

When a comment is not trailing to the class definition but actually to one of the statements in the parent suite:

```python
if top1:
    class Nested:
        pass
# comment
if top2:
    pass
```

#### 3

Similar to the previous case but at end of file:

```python
if top1:
    class Nested:
        pass
# comment
```

### Number of empty lines

We also need to make sure that running the formatter multiple times doesn't keep on adding empty lines (at max it could go to 2). This would happen when the new logic would add an empty line on the first run and then on the second run another logic sees the empty line which it then tries to maintain. This adds multiple empty line IR elements.

## Possible solutions

Instead of looking ahead, we'll always look behind which is to say that for the current node we need to query the absolute previous node. But we'll have to add the logic in multiple places - suite, if, match, try, etc. Here, absolute previous in the literal sense, not just in the same indent level. For example,

```
class Top:
    class Nested11:
        class Nested12:
            pass
    class Nested21:
        pass
```

Here, the previous statement for `Nested21` would be `Nested12` and not `Nested11`.


[the loop]: https://github.com/astral-sh/ruff/blob/86a5375e04b296ca3b6e1717ee07009cf5977930/crates/ruff_python_formatter/src/statement/suite.rs#L190

File: crates/ruff_python_formatter/src/statement/suite.rs (hunk 2 above)
Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/statement/suite.rs]: Yes

File: crates/ruff_python_formatter/src/comments/format.rs [hunk 3]
```diff
@@ -85,45 +86,61 @@ pub(crate) struct FormatLeadingAlternateBranchComments<'a> {
...
 impl Format<PyFormatContext<'_>> for FormatLeadingAlternateBranchComments<'_> {
     fn fmt(&self, f: &mut PyFormatter) -> FormatResult<()> {
+        let empty_line_condition = if is_blank_line_after_nested_stub_class_enabled(f.context()) {
+            self.last_node.map_or(false, |preceding| {
+                blank_line_after_nested_stub_class_condition(preceding, None, f)
+            })
+        } else {
+            false
+        };
+
```
Comment (dhruvmanila) [inline crates/ruff_python_formatter/src/comments/format.rs:96]: This helps when the class is at the boundary location in nested statement:

```python
if foo:
    class Nested:
        pass
else:
    pass
```

The `FormatLeadingAlternateBranchComments` will be called before the `else` block formatting. Similarly for other blocks like `match`, `case`, `try .. except ..`, etc.

File: crates/ruff_python_formatter/src/comments/format.rs [hunk 4]
```diff
@@ -513,14 +530,32 @@ fn strip_comment_prefix(comment_text: &str) -> FormatResult<&str> {
...
+    node_kind: NodeKind,
 ) -> FormatEmptyLinesBeforeTrailingComments<'a> {
     // Black has different rules for stub vs. non-stub and top level vs. indented
     let empty_lines = match (f.options().source_type(), f.context().node_level()) {
         (PySourceType::Stub, NodeLevel::TopLevel(_)) => 1,
-        (PySourceType::Stub, _) => 0,
+        (PySourceType::Stub, _) => u32::from(
+            is_blank_line_after_nested_stub_class_enabled(f.context())
+                && node_kind == NodeKind::StmtClassDef,
+        ),
```
Comment (dhruvmanila) [inline crates/ruff_python_formatter/src/comments/format.rs:558]: Logic for when a class is followed by a trailing comment.

File: crates/ruff_python_formatter/src/statement/suite.rs [hunk 5]
```diff
@@ -472,15 +472,28 @@ fn stub_file_empty_lines(
...
-            if empty_line_condition
-                && lines_after_ignoring_end_of_line_trivia(preceding.end(), source) > 1
+            if (empty_line_condition
+                && lines_after_ignoring_end_of_line_trivia(preceding.end(), source) > 1)
+                || (is_blank_line_after_nested_stub_class_enabled(f.context())
+                    && blank_line_after_nested_stub_class_condition(
+                        preceding.into(),
+                        Some(following.into()),
+                        f,
+                    ))
```
Comment (dhruvmanila) [inline crates/ruff_python_formatter/src/statement/suite.rs:496]: It's difficult to merge the `is_*_enabled && *_condition` logic with `empty_line_condition` variable up above like so:

```rust
    let empty_line_condition = preceding_comments.has_trailing()
        || following_comments.has_leading()
        || !stub_suite_can_omit_empty_line(preceding, following, f)
        || (is_blank_line_after_nested_stub_class_enabled(f.context())
                    && blank_line_after_nested_stub_class_condition(
                        preceding.into(),
                        Some(following.into()),
                        f,
                    ))
```

The reason being that in the top level condition the boolean check is direct but in other suite kind there's an additional check of the number of empty lines. That additional check is to _only_ preserve the empty lines if they're present while the preview style adds an empty line conditionally. So, the condition should OR the complete check which was existing earlier to complement it.

File: crates/ruff_python_formatter/src/statement/suite.rs [hunk 6]
```diff
@@ -490,6 +503,40 @@ fn stub_file_empty_lines(
...
+/// Checks if an empty line should be inserted between the preceding and, optionally,
+/// the following node according to the [`blank_line_after_nested_stub_class`](https://github.com/astral-sh/ruff/issues/8891)
+/// preview style.
+///
+/// If `following` is `None`, then the preceding node is the last node in the suite.
+pub(crate) fn blank_line_after_nested_stub_class_condition(
+    preceding: AnyNodeRef<'_>,
+    following: Option<AnyNodeRef<'_>>,
+    f: &PyFormatter,
+) -> bool {
```
Comment (dhruvmanila) [inline crates/ruff_python_formatter/src/statement/suite.rs:515]: This is the core condition which checks whether we need to add an empty line between the preceding and following node or not.

File: crates/ruff_python_formatter/src/statement/suite.rs [hunk 7]
```diff
@@ -490,6 +503,40 @@ fn stub_file_empty_lines(
...
+                    Some(AnyNodeRef::StmtClassDef(ast::StmtClassDef {
+                        body,
+                        decorator_list,
+                        ..
+                    })) => !contains_only_an_ellipsis(body, comments) || !decorator_list.is_empty(),
+                    Some(AnyNodeRef::StmtFunctionDef(_)) | None => true,
+                    _ => false,
+                }
+        }
+        Some(_) => !comments.has_trailing_own_line(preceding),
```
Comment (dhruvmanila) [inline crates/ruff_python_formatter/src/statement/suite.rs:530]: Class with a trailing comment is handled by `empty_lines_before_trailing_comments` builder.

File: crates/ruff_python_formatter/src/statement/suite.rs [hunk 8]
```diff
@@ -490,6 +503,40 @@ fn stub_file_empty_lines(
...
+        None => std::iter::successors(
+            preceding.last_child_in_body(),
+            AnyNodeRef::last_child_in_body,
+        )
+        .take_while(|last_child| !comments.has_trailing_own_line(*last_child))
+        .any(|last_child| last_child.is_stmt_class_def()),
```
Comment (dhruvmanila) [inline crates/ruff_python_formatter/src/statement/suite.rs:536]: If the preceding node isn't a class, then we check if there's a class node as the last child node of the body of nested nodes. This nesting could be arbitrarily deep.

File: crates/ruff_python_formatter/src/statement/suite.rs [hunk 9]
```diff
@@ -472,15 +472,28 @@ fn stub_file_empty_lines(
...
         SuiteKind::TopLevel => {
-            if empty_line_condition {
+            if empty_line_condition
+                || (is_blank_line_after_nested_stub_class_enabled(f.context())
+                    && blank_line_after_nested_stub_class_condition(
+                        preceding.into(),
+                        Some(following.into()),
+                        f,
+                    ))
+            {
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/statement/suite.rs:482]: Moving the `is_blank_line_after_nested_stub_class_enabled` into `blank_lines_after_nested_stub_class_condition` simplifies the calling code a lot

File: crates/ruff_python_formatter/src/statement/suite.rs (hunk 5 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/statement/suite.rs:496]: I think what would help here is to rename `empty_line_condition` into `preserve_empty_line` which is different from your check which forces empty lines. 

I recommend moving the `blank_line_after_nested_stub_class_condition` out of the `match` and next to `empty_line_condition`, and name the variable `require_empty_line`

I would rename `blank_line_after_nested_stub_class_condition` to `requires_empty_line_between_stub_file_statements` (not just preserving, enforcing an empty line).

We can then rewrite the checks in the match branches to 

```
if requires_empty_line || preserve_empty_line {
    empty_line.fmt(f)
}
```

which I find easier to understand.

File: crates/ruff_python_formatter/src/statement/suite.rs [hunk 10]
```diff
@@ -490,6 +503,40 @@ fn stub_file_empty_lines(
...
+            !class.decorator_list.is_empty()
+                || match following {
+                    Some(AnyNodeRef::StmtClassDef(ast::StmtClassDef {
+                        body,
+                        decorator_list,
+                        ..
+                    })) => !contains_only_an_ellipsis(body, comments) || !decorator_list.is_empty(),
+                    Some(AnyNodeRef::StmtFunctionDef(_)) | None => true,
+                    _ => false,
+                }
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/statement/suite.rs:528]: Nit: A match in a binary expression feels very heavy. Some temporary variables could also help to document the conditions. 
```suggestion
            let has_decorators = !class.decorator_list.is_empty();
            let followed_by_class_or_function = match following {
                Some(AnyNodeRef::StmtClassDef(ast::StmtClassDef {
                    body,
                    decorator_list,
                    ..
                })) => !contains_only_an_ellipsis(body, comments) || !decorator_list.is_empty(),
                Some(AnyNodeRef::StmtFunctionDef(_)) | None => true,
                _ => false,
            };

            has_decorators || followed_by_class_or_function
```

I think it would be helpful to add some comments here explaining why the conditions are as they are.

**Review by MichaReiser — COMMENTED** [review]
The code changes overall look good to me. I think we can make some improvements to improve readability and extend documentation a bit.


I'm very surprised that there are no changed black snapshots. Do you know why that is? Are we sure we match Black's formatting?



It would help reviewers if you add an example of the expected formatting (how is it different from today) with a short explanation of the preview style to the PR. I know it's in the linked PR, but it requires multiple steps by the reviewer: Open issue, read the summary, go back to the PR, read the summary, and finally read the code.

**Comment by MichaReiser** [comment]
#9674 should ensure that black stub files with options run as stub tests.

File: crates/ruff_python_formatter/src/comments/format.rs [hunk 11]
```diff
@@ -85,45 +86,61 @@ pub(crate) struct FormatLeadingAlternateBranchComments<'a> {
 
 impl Format<PyFormatContext<'_>> for FormatLeadingAlternateBranchComments<'_> {
     fn fmt(&self, f: &mut PyFormatter) -> FormatResult<()> {
+        let empty_line_condition = if is_blank_line_after_nested_stub_class_enabled(f.context()) {
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/comments/format.rs:89]: Can you explain why this special handling is required here but not for other cases for which `stub_suite_can_omit_empty_line` returns false?

File: crates/ruff_python_formatter/src/comments/format.rs (hunk 4 above)
Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/comments/format.rs:558]: I did update the docs for the function :)

https://github.com/astral-sh/ruff/blob/a768fc35ca1f73012ae9d76cba2eedda95c9db1e/crates/ruff_python_formatter/src/comments/format.rs#L534-L546

Is this what you meant?

File: crates/ruff_python_formatter/src/statement/suite.rs (hunk 5 above)
Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/statement/suite.rs:496]: > I think what would help here is to rename `empty_line_condition` into `preserve_empty_line` which is different from your check which forces empty lines.

I'm not sure if this is 100% correct. For top-level, if the condition is true, we always make sure that there is just one empty line. It could be that there isn't any empty line in the source code which means we're adding an additional empty line. But, for the nested cases, we'd only preserve any existing empty lines and limit it to 1.

> I recommend moving the `blank_line_after_nested_stub_class_condition` out of the `match` and next to `empty_line_condition`, and name the variable `require_empty_line`

Thanks, I'll update the code accordingly.

> I would rename `blank_line_after_nested_stub_class_condition` to `requires_empty_line_between_stub_file_statements` (not just preserving, enforcing an empty line).

Yeah, I'm going with `is_blank_line_required_after_nested_stub_class`.

Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/statement/suite.rs:496]: > Yeah, I'm going with `is_blank_line_required_after_nested_stub_class`.

Actually, with your other comment, I'm thinking of `should_insert_blank_line_after_class_in_stub_file`.

File: crates/ruff_python_formatter/src/statement/suite.rs [hunk 12]
```diff
@@ -490,6 +503,40 @@ fn stub_file_empty_lines(
     }
 }
 
+/// Checks if an empty line should be inserted between the preceding and, optionally,
+/// the following node according to the [`blank_line_after_nested_stub_class`](https://github.com/astral-sh/ruff/issues/8891)
+/// preview style.
+///
+/// If `following` is `None`, then the preceding node is the last node in the suite.
+pub(crate) fn blank_line_after_nested_stub_class_condition(
```
[the remark this answers is not in the document]
Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/statement/suite.rs:511]: Good point. It isn't specific to nested class. I'll update the name.

File: crates/ruff_python_formatter/src/comments/format.rs (hunk 11 above)
Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/comments/format.rs]: This is to handle cases when a class definition is the last statement in a suite. The empty line formatting in suite is a loop which handles only for in between statements i.e., both preceding and following are present. This is required only before an alternate branch because otherwise it'll be handled by the top-level suite formatting.

File: crates/ruff_python_formatter/src/statement/suite.rs (hunk 10 above)
Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/statement/suite.rs]: Thanks for the suggestion. I've added some comments with examples.

File: crates/ruff_python_formatter/src/comments/format.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/comments/format.rs]: Yes, thanks for noticing that.

File: crates/ruff_python_formatter/src/statement/suite.rs (hunk 9 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/statement/suite.rs]: Yeah that helps. It might still be worth to move the check into the "condition" function because the check is necessary at all call-sites of that function (and leads to very long conditions)
